pkg/natdiscovery/listener.go

```diff
@@ -17,42 +17,54 @@ limitations under the License.
 package natdiscovery
 
 import (
+	"encoding/hex"
 	"net"
+	"os"
 	"strconv"
 	"time"
 
 	"github.com/pkg/errors"
 	"google.golang.org/protobuf/proto"
-	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/klog"
 
 	natproto "github.com/submariner-io/submariner/pkg/natdiscovery/proto"
 )
 
 func (nd *natDiscovery) runListener(stopCh <-chan struct{}) error {
-	serverAddress, err := net.ResolveUDPAddr("udp4", ":"+strconv.Itoa(natproto.DefaultPort))
+	serverAddress, err := net.ResolveUDPAddr("udp4", ":"+strconv.Itoa(int(*nd.localEndpoint.Spec.NATDiscoveryPort)))
 	if err != nil {
 		return errors.Wrap(err, "Error resolving UDP address")
 	}
 
-	serverConnection, err := net.ListenUDP("udp", serverAddress)
+	serverConnection, err := net.ListenUDP("udp4", serverAddress)
 	if err != nil {
-		return errors.Wrapf(err, "Error listening on udp port %d", natproto.DefaultPort)
+		return errors.Wrapf(err, "Error listening on udp port %d", *nd.localEndpoint.Spec.NATDiscoveryPort)
 	}
 
 	// Instead of storing the server connection I save the reference to the WriteToUDP
 	// of our server connection instance, in a way that we can use this for unit testing
 	// later too.
 	nd.serverUDPWrite = serverConnection.WriteToUDP
 
-	go wait.Until(func() {
-		buf := make([]byte, 2048)
-		if _, addr, err := serverConnection.ReadFromUDP(buf); err != nil {
-			klog.Errorf("Error receiving from udp: %s", err)
-		} else if err := nd.parseAndHandleMessageFromAddress(buf, addr); err != nil {
-			klog.Errorf("Error handling message from address %#v: %s", addr, err)
+	buf := make([]byte, 2048)
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				serverConnection.Close()
+				return
+			default:
+				serverConnection.SetReadDeadline(time.Now().Add(10 * time.Second))
```

Comment:
This should probably be much lower.

Reply:
Yup, let's wait on a separate goroutine and close the socket, that makes sense and it's cleaner.